pkg/admission/customresourcevalidation/customresourcevalidator.go

```diff
@@ -0,0 +1,114 @@
+	if name != "cluster" {
+		return []string{"must be cluster"}
+	}
+	return nil
+}
+
+type ValidateObjCreateFunc func(obj runtime.Object) field.ErrorList
+
+type ValidateObjUpdateFunc func(obj runtime.Object, oldObj runtime.Object) field.ErrorList
+
+type ValidateStatusUpdateFunc func(obj runtime.Object, oldObj runtime.Object) field.ErrorList
+
+// ValidateCustomResource is an implementation of admission.Interface.
+// It looks at all new pods and overrides each container's image pull policy to Always.
+type validateCustomResource struct {
+	*admission.Handler
+
+	resource               schema.GroupResource
+	validateObjCreateFn    ValidateObjCreateFunc
+	validateObjUpdateFn    ValidateObjUpdateFunc
+	validateStatusUpdateFn ValidateStatusUpdateFunc
+}
+
+func NewValidator(resource schema.GroupResource, validateObjCreateFn ValidateObjCreateFunc, validateObjUpdateFn ValidateObjUpdateFunc, validateStatusUpdateFn ValidateStatusUpdateFunc) (admission.Interface, error) {
+	if validateObjCreateFn == nil {
+		return nil, fmt.Errorf("createFn is required")
+	}
+	if validateObjUpdateFn == nil {
+		return nil, fmt.Errorf("updateFn is required")
+	}
+	if validateStatusUpdateFn == nil {
+		return nil, fmt.Errorf("updateFn is required")
+	}
+
+	return &validateCustomResource{
+		Handler:                admission.NewHandler(admission.Create, admission.Update),
+		resource:               resource,
+		validateObjCreateFn:    validateObjCreateFn,
+		validateObjUpdateFn:    validateObjUpdateFn,
+		validateStatusUpdateFn: validateStatusUpdateFn,
+	}, nil
+}
+
+var _ admission.ValidationInterface = &validateCustomResource{}
+
+func (a *validateCustomResource) Validate(uncastAttributes admission.Attributes) error {
+	attributes := &unstructuredUnpackingAttributes{Attributes: uncastAttributes}
+	if a.shouldIgnore(attributes) {
+		return nil
+	}
+
+	switch attributes.GetOperation() {
+	case admission.Create:
+		// creating subresources isn't something we understand, but we can be pretty sure we don't need to validate it
+		if len(attributes.GetSubresource()) > 0 {
+			return nil
+		}
+		errors := a.validateObjCreateFn(attributes.GetObject())
+		if len(errors) == 0 {
+			return nil
+		}
+		return apierrors.NewInvalid(attributes.GetKind().GroupKind(), attributes.GetName(), errors)
+
+	case admission.Update:
+		switch attributes.GetSubresource() {
+		case "":
+			errors := a.validateObjUpdateFn(attributes.GetObject(), attributes.GetOldObject())
+			if len(errors) == 0 {
+				return nil
+			}
+			return apierrors.NewInvalid(attributes.GetKind().GroupKind(), attributes.GetName(), errors)
+
+		case "status":
+			errors := a.validateStatusUpdateFn(attributes.GetObject(), attributes.GetOldObject())
+			if len(errors) == 0 {
+				return nil
+			}
+			return apierrors.NewInvalid(attributes.GetKind().GroupKind(), attributes.GetName(), errors)
+
+		default:
+			admission.NewForbidden(attributes, fmt.Errorf("unhandled subresource: %v", attributes.GetSubresource()))
```

Comment:
Should this error (and the other Forbidden error below) not be returned?

Reply:
> Should this error (and the other Forbidden error below) not be returned?

bug.

Reply:
For this specific line, should we not just ignore (i.e. `return nil`), to be consistent with `shouldIgnore()`?